### ISSUE TITLE
GUI Module key callbacks

### ISSUE THREAD

**Issue by cygann** [intel-isl/Open3D#2320]
Does the new GUI Module have support for registering key callbacks like the open3d `visualization.VisualizerWithKeyCallback` class? If so, where should I look?

Thanks!

**Comment by Zumbalamambo** [comment]
@cygann Yes it is possible... I wrote a minimalistic example for you. 

```

from functools import partial

import open3d as o3d
import numpy as np

path = "file_name"


def save_image(vis, type):
    fpath = "{}_{}.jpg".format(path, type)
    vis.capture_screen_image(fpath)
    print("{} was saved.".format(fpath))
    return False


pcd = o3d.io.read_point_cloud("assets/pointcloud.pcd")
vis = o3d.visualization.VisualizerWithKeyCallback()
vis.create_window()
vis.get_render_option().background_color = np.asarray([0, 0, 0])

view_ctl = vis.get_view_control()
vis.add_geometry(pcd)
vis.register_key_callback(ord("S"), partial(save_image, type="pcd"))
vis.run()
vis.destroy_window()
```


Press `S` and the `save_image` method will be invoked. Try this and let us know if it has any issues.

**Comment by Zumbalamambo** [comment]
@cygann Yes it is possible...

```

from functools import partial

import open3d as o3d
import numpy as np

path = "file_name"


def save_image(vis, type):
    fpath = "{}_{}.jpg".format(path, type)
    vis.capture_screen_image(fpath)
    print("{} was saved.".format(fpath))
    return False


pcd = o3d.io.read_point_cloud("assets/pointcloud.pcd")
vis = o3d.visualization.VisualizerWithKeyCallback()
vis.create_window()
vis.get_render_option().background_color = np.asarray([0, 0, 0])

view_ctl = vis.get_view_control()
vis.add_geometry(pcd)
vis.register_key_callback(ord("S"), partial(save_image, type="pcd"))
vis.run()
vis.destroy_window()
```


Press `S` and the `save_image` method will be invoked. Try this and let us know if it has any issues.

**Comment by Zumbalamambo** [comment]
Yeah it is possible... Try this code,

```

from functools import partial

import open3d as o3d
import numpy as np

path = "file_name.png"


def save_image(vis):
    vis.capture_screen_image(path)
    print("Saved")
    return False


pcd = o3d.io.read_point_cloud("assets/pointcloud.pcd")
vis = o3d.visualization.VisualizerWithKeyCallback()
vis.create_window()
vis.get_render_option().background_color = np.asarray([0, 0, 0])

view_ctl = vis.get_view_control()
vis.add_geometry(pcd)
vis.register_key_callback(ord("S"), partial(save_image))
vis.run()
vis.destroy_window()
```

Press `S` and it will execute the `save_image`. In addition, you could also pass additional arguments if you want to with minor modification.

**Comment by cygann** [comment]
I'm familiar with how to use the VisualizerWithKeyCallback class, but I'm not sure that it's compatible with the new GUI module since it uses its own visualizer. I'm looking to use the callbacks in a program similar to [this one](https://github.com/intel-isl/Open3D/blob/master/examples/python/GUI/vis-gui.py)

**Comment by hduonggithub** [comment]
I have the same question too.

**Comment by prewettg** [comment]
The GUI module does not have facilities for registering key callbacks at the moment, but it is easy to add a menu item or button, both of which have callbacks when triggered, would that work for your use case?

**Comment by hduonggithub** [comment]
@prewettg 

I am eager to know how to do that. Can you please give a hint or example code to do that. Python or C++ is fine. Thanks!

**Comment by prewettg** [comment]
Check out the examples in `examples/python/gui`. [add-geometry.py](https://github.com/intel-isl/Open3D/blob/master/examples/python/gui/add-geometry.py) is a simple example with menu items, and [vis-gui.py](https://github.com/intel-isl/Open3D/blob/master/examples/python/gui/vis-gui.py) is essentially a python implementation of the viewer app, so it is pretty full-featured, but much less simple. One thing I should note is that right now a SceneWidget always draws on the bottom of the z-order, which means you cannot put it in a layout (the layout draws on top of it). So you will need to call `window.set_on_layout()` to register a callback that runs when the window resizes, and then set the sizes of all the top-level children manually (`vis-gui.py` demonstrates this).

There is not a C++ example per se, but [cpp/open3d/visualization/visualizer/GuiVisualizer.cpp](https://github.com/intel-isl/Open3D/blob/master/cpp/open3d/visualization/visualizer/GuiVisualizer.cpp) is the implementation of the viewer app. The python bindings in [cpp/pybind/visualization/gui/gui.cpp](https://github.com/intel-isl/Open3D/blob/master/cpp/pybind/visualization/gui/gui.cpp) may also be useful.

**Comment by prewettg** [comment]
You can now add mouse/key event callbacks to SceneWidget in Python. (In C++ you can always just inherit from SceneWidget and override Mouse() and/or Key()).  This should be available in 0.12.  Here is a simple example for using them:
```
#!/usr/bin/env python
import open3d as o3d
import open3d.visualization.gui as gui

def on_mouse(e):
    if e.type == gui.MouseEvent.Type.BUTTON_DOWN:
        print("[debug] mouse:", (e.x, e.y))
    return gui.Widget.EventCallbackResult.IGNORED

def on_key(e):
    if e.key == gui.KeyName.SPACE:
        if e.type == gui.KeyEvent.UP:  # check UP so we default to DOWN
            print("[debug] SPACE released")
        else:
            print("[debug] SPACE pressed")
        return gui.Widget.EventCallbackResult.HANDLED
    if e.key == gui.KeyName.W:  # eats W, which is forward in fly mode
        print("[debug] Eating W")
        return gui.Widget.EventCallbackResult.CONSUMED
    return gui.Widget.EventCallbackResult.IGNORED

def main():
    gui.Application.instance.initialize()
    w = gui.Application.instance.create_window("Open3D Example - Events",
                                               640, 480)
    scene = gui.SceneWidget()
    scene.scene = o3d.visualization.rendering.Open3DScene(w.renderer)
    w.add_child(scene)

    obj = o3d.geometry.TriangleMesh.create_moebius()
    obj.compute_vertex_normals()

    material = o3d.visualization.rendering.Material()
    material.shader = "defaultLit"
    scene.scene.add_geometry("Moebius", obj, material)

    scene.setup_camera(60, scene.scene.bounding_box, (0, 0, 0))
    # scene.set_view_controls(gui.SceneWidget.Controls.FLY)
    scene.set_on_mouse(on_mouse)
    scene.set_on_key(on_key)

    gui.Application.instance.run()

if __name__ == "__main__":
    main()
```
